src/elf.c

```diff
@@ -363,42 +546,263 @@ void elf_generate_sections(void)
+                      LIBC_SO); /* Add "libc.so.6" to .dynstr. */
+        elf_write_byte(dynamic_sections.elf_dynstr, 0);
+        st_name += strlen(LIBC_SO) + 1;
+
+        /* Add __libc_start_main explicitly. */
+        elf_write_str(dynamic_sections.elf_dynstr, "__libc_start_main");
+        elf_write_byte(dynamic_sections.elf_dynstr, 0);
+        sym.st_name = st_name;
+        sym.st_info = ELF32_ST_INFO(1, 2); /* STB_GLOBAL = 1, STT_FUNC = 2 */
+        elf_write_blk(dynamic_sections.elf_dynsym, &sym, sizeof(elf32_sym_t));
+        st_name += strlen("__libc_start_main") + 1;
+        rel.r_offset = dynamic_sections.elf_got_start + PTR_SIZE * 3;
+        rel.r_info = (dymsym_idx << 8) | R_ARCH_JUMP_SLOT;
+        elf_write_blk(dynamic_sections.elf_relplt, &rel, sizeof(elf32_rel_t));
+        dymsym_idx += 1;
+        func_plt_ofs = PLT_FIXUP_SIZE + PLT_ENT_SIZE;
+        func_got_ofs = PTR_SIZE << 2;
+        for (func_t *func = FUNC_LIST.head; func; func = func->next) {
+            if (func->is_used && !func->bbs) {
+                /* Handle all functions that need relocation */
+                elf_write_str(dynamic_sections.elf_dynstr,
+                              func->return_def.var_name);
+                elf_write_byte(dynamic_sections.elf_dynstr, 0);
+                sym.st_name = st_name;
+                sym.st_info = ELF32_ST_INFO(1, 2);
+                elf_write_blk(dynamic_sections.elf_dynsym, &sym,
+                              sizeof(elf32_sym_t));
+                st_name += strlen(func->return_def.var_name) + 1;
+                rel.r_offset += PTR_SIZE;
+                rel.r_info = (dymsym_idx << 8) | R_ARCH_JUMP_SLOT;
+                elf_write_blk(dynamic_sections.elf_relplt, &rel,
+                              sizeof(elf32_rel_t));
+                dymsym_idx += 1;
+                func->plt_offset = func_plt_ofs;
+                func->got_offset = func_got_ofs;
+                func_plt_ofs += PLT_ENT_SIZE;
+                func_got_ofs += PTR_SIZE;
+            }
+        }
+        elf_align(dynamic_sections.elf_dynstr);
+
+        /* got section */
+        elf_write_int(dynamic_sections.elf_got,
+                      dynamic_sections.elf_got_start +
+                          dynamic_sections.got_size +
+                          dynamic_sections.elf_dynstr->size +
+                          dynamic_sections.elf_dynsym
+                              ->size); /* The virtual address of .dynamic. */
+        elf_write_int(dynamic_sections.elf_got,
+                      0); /* Set got[1] to 0 for link_map. */
+        elf_write_int(dynamic_sections.elf_got,
+                      0); /* Set got[2] to 0 for resolver. */
+        for (int i = PTR_SIZE * 3; i < dynamic_sections.got_size - PTR_SIZE;
+             i += PTR_SIZE)
+            elf_write_int(dynamic_sections.elf_got,
+                          dynamic_sections.elf_plt_start);
+        elf_write_int(dynamic_sections.elf_got, 0); /* End with 0x0 */
```

Comment:
Why is `GOT[N] = 0` needed as an end marker?

Dynamic linker determines GOT size via `DT_PLTGOT` and `DT_PLTRELSZ`, does not need a terminator.